File: src/main/java/features/repositoryaccessors/RepositoryAccessorBlock.java
package features.repositoryaccessors;

import core.algorithm.Algorithm;
import core.data.Data;
import core.pipeline.PipelineBlock;
import core.repository.RepositoryCommander;
import features.repositoryaccessors.data.RepoRequest;

/**
 * {@link PipelineBlock} that stores all necessary information about repository such as {@link RepositoryCommander},
 * connection string and Repository accessor algorithm
 * @param <I> Input type
 * @param <O> Output type
 */
public class RepositoryAccessorBlock<I extends Data, O extends Data> extends PipelineBlock<I, O> {

    private RepositoryCommander repo;
    private Algorithm<RepoRequest, O> algorithm;

    public boolean enabled = true;
    private String connectionString;

    public RepositoryAccessorBlock(RepositoryCommander repo, Algorithm<RepoRequest, O> algorithm, String connectionString) {
        this.repo = repo;
        this.algorithm = algorithm;
        this.connectionString = connectionString;
    }

    @Override
    public void inputDataReady(PipelineBlock<?, I> sender, I input) {
        if(enabled) {
            O result = algorithm.execute(new RepoRequest(connectionString, repo){{bundle = input;}});
            notifyListeners(this, result);
        }
    }
}

File: src/main/java/features/repositoryaccessors/RepositoryAccessorBlock.kt
package features.repositoryaccessors

import core.algorithm.Algorithm
import core.data.Data
import core.pipeline.DummyPipeKeeper
import core.pipeline.PipelineBlock
import core.pipeline.PipelineException
import core.pipeline.PipelineKeeper
import core.repository.RepositoryCommander
import features.repositoryaccessors.data.RepoRequest

/**
 * [PipelineBlock] that stores all necessary information about repository such as [RepositoryCommander],
 * connection string and Repository accessor algorithm
 * @param <I> Input type
 * @param <O> Output type
</O></I> */
class RepositoryAccessorBlock<I : Data, O : Data>(
        private val repo: RepositoryCommander,
        private val algorithm: Algorithm<RepoRequest, O>,
        private val connectionString: String,
        name: String = "Repository accessor for $algorithm",
        pipelineKeeper: PipelineKeeper = DummyPipeKeeper()) : PipelineBlock<I, O>(name, pipelineKeeper){

    var enabled = true

    override fun inputReady(sender: Any, input: I) {
        if (enabled) {

            val record = pipelineKeeper.session.addNew("[$name]: request")

            try {
                val result = algorithm.execute(RepoRequest(connectionString, repo, input))
                onDataReady(this, result)
                record.setSuccess()
            } catch (e: PipelineException) {
                record.setError()
                throw e
            }
        }
    }

    override fun flush() { }
}
